File: modules/rag-retriever.js
/**
 * RAG Retriever
 * Handles query processing and document retrieval for RAG
 */
const path = require('path');
const fs = require('fs-extra');

class RagRetriever {
  /**
   * Create a new RAG retriever
   * @param {Object} vectorDB - Vector database adapter
   * @param {Object} options - Configuration options
   */
  constructor(vectorDB, options = {}) {
    this.vectorDB = vectorDB;
    this.options = Object.assign({
      topK: 5,
      reranking: true,
      minScore: 0.7,
      keywordBoost: true,
      recencyBoost: true,
      chunkMerging: true,
      ragDir: 'rag_knowledge_base',
      encoder: null
    }, options);
    
    this.encoder = this.options.encoder;
    this.initialized = false;
  }
  
  /**
   * Initialize the retriever
   */
  async initialize() {
    if (this.initialized) return;
    
    try {
      // Load encoder model if needed
      if (!this.encoder) {
        try {
          // First check if we should use the MLAnalyzer
          if (this.options.mlAnalyzer && 
              typeof this.options.mlAnalyzer.generateDocumentEmbedding === 'function') {
            console.log("Using MLAnalyzer for embedding generation");
            this.encoder = {
              embed: async (text) => {
                const embedding = await this.options.mlAnalyzer.generateDocumentEmbedding(text);
                return { array: async () => [embedding] };
              }
            };
          } else {
            // Try to load TensorFlow.js first
            console.log("Loading TensorFlow.js...");
            try {
              // Try to load tfjs-node first (faster)
              require('@tensorflow/tfjs-node');
            } catch (err) {
              console.log("TensorFlow.js Node not available, using browser version");
              require('@tensorflow/tfjs');
            }
            
            console.log("Loading Universal Sentence Encoder...");
            this.encoder = await require('@tensorflow-models/universal-sentence-encoder').load();
            console.log("Universal Sentence Encoder loaded successfully");
          }
        } catch (err) {
          console.error("Error loading encoder for RAG:", err.message);
          console.log("Using fallback embedding method");
          // Very simple fallback embedding (not for production use)
          this.encoder = {
            embed: async (text) => {
              // Simple hash function to generate pseudo-embeddings
              const hash = [...text].reduce((acc, char) => {
                return ((acc << 5) - acc) + char.charCodeAt(0);
              }, 0);
              
              // Generate a deterministic but simple vector based on the hash
              const vector = new Array(512).fill(0).map((_, i) => {
                return Math.sin(hash * (i + 1) * 0.01) * 0.5 + 0.5;
              });
              
              return { array: async () => [vector] };
            }
          };
        }
      }
      
      this.initialized = true;
    } catch (err) {
      console.error("Failed to initialize RAG retriever:", err.message);
      throw err;
    }
  }
  
  /**
   * Generate embedding for query text
   * @param {string} query - Query text
   * @returns {Array<number>} Embedding vector
   */
  async generateQueryEmbedding(query) {
    try {
      const embedding = await this.encoder.embed(query);
      const embeddingArray = await embedding.array();
      return embeddingArray[0];
    } catch (err) {
      console.error("Error generating query embedding:", err.message);
      return null;
    }
  }
  
  /**
   * Retrieve relevant documents for a query
   * @param {string} query - User query
   * @param {Object} options - Query options
   * @returns {Array<Object>} Relevant documents
   */
  async retrieveDocuments(query, options = {}) {
    await this.initialize();
    
    const queryOptions = { ...this.options, ...options };
    
    console.log(`Retrieving documents for query: "${query}"`);
    
    // Generate query embedding
    const queryEmbedding = await this.generateQueryEmbedding(query);
    if (!queryEmbedding) {
      throw new Error("Failed to generate query embedding");
    }
    
    // Apply filters
    const filters = this.buildFilters(queryOptions.filters);
    
    // Retrieve similar documents
    const results = await this.vectorDB.findSimilar(
      queryEmbedding, 
      {
        topK: queryOptions.topK * (queryOptions.reranking ? 2 : 1), // Get more results for reranking
        filter: filters
      }
    );
    
    if (results.length === 0) {
      console.log("No documents found for query");
      return [];
    }
    
    console.log(`Found ${results.length} initial matches`);
    
    // Rerank results if enabled
    const finalResults = queryOptions.reranking ? 
      await this.rerankResults(query, results, queryOptions) : 
      results;
    
    // Filter by minimum score
    const filteredResults = finalResults.filter(
      result => result.score >= queryOptions.minScore
    );
    
    // Fetch document content if not included in results
    const resultsWithContent = await this.fetchDocumentContent(filteredResults);
    
    // Merge chunks from the same document if enabled
    const mergedResults = queryOptions.chunkMerging ? 
      this.mergeDocumentChunks(resultsWithContent) : 
      resultsWithContent;
    
    console.log(`Retrieved ${mergedResults.length} relevant documents`);
    return mergedResults;
  }
  
  /**
   * Build filter expression
   * @param {Object} filters - Filter criteria
   * @returns {Object} Database-specific filters
   */
  buildFilters(filters = {}) {
    if (!filters || Object.keys(filters).length === 0) return {};
    
    // Convert simple filters object to database-specific format
    const dbFilters = {};
    
    // Handle system filter specifically
    if (filters.system) {
      dbFilters["system"] = Array.isArray(filters.system) ? 
        filters.system : [filters.system];
    }
    
    // Handle category filter
    if (filters.category) {
      dbFilters["category"] = Array.isArray(filters.category) ? 
        filters.category : [filters.category];
    }
    
    // Handle other filters
    for (const [key, value] of Object.entries(filters)) {
      if (key !== 'system' && key !== 'category') {
        dbFilters[key] = value;
      }
    }
    
    return dbFilters;
  }
  
  /**
   * Rerank results using additional heuristics
   * @param {string} query - Original query
   * @param {Array<Object>} results - Initial results
   * @param {Object} options - Reranking options
   * @returns {Array<Object>} Reranked results
   */
  async rerankResults(query, results, options) {
    // Extract keywords from query
    const keywords = this.extractKeywords(query);
    
    // Score results with hybrid approach
    const scoredResults = results.map(result => {
      let finalScore = result.score; // Start with vector similarity score
      
      // Apply keyword boosting if enabled
      if (options.keywordBoost) {
        let keywordBoost = 0;
        
        // Check how many keywords match in the content
        const resultText = JSON.stringify(result.metadata).toLowerCase();
        keywords.forEach(keyword => {
          if (resultText.includes(keyword.toLowerCase())) {
            keywordBoost += 0.02; // Small boost per keyword match
          }
        });
        
        finalScore += keywordBoost;
      }
      
      // Apply recency boost if enabled and lastUpdated is available
      if (options.recencyBoost && result.metadata.lastUpdated) {
        const age = (Date.now() - new Date(result.metadata.lastUpdated).getTime()) / (1000 * 60 * 60 * 24); // Age in days
        const recencyBoost = Math.max(0, 0.05 * (1 - Math.min(age / 30, 1))); // Up to 0.05 boost for recent docs
        finalScore += recencyBoost;
      }
      
      // Boost system-specific content if specified in query
      const systemMatches = {
        canvas: /canvas|lms|course|learning/i,
        discourse: /discourse|forum|topic|discussion/i,
        integration: /integrat|connect|combine/i
      };
      
      for (const [system, pattern] of Object.entries(systemMatches)) {
        if (pattern.test(query) && result.metadata.system === system) {
          finalScore += 0.03; // Boost for matching system context
        }
      }
      
      return {
        ...result,
        originalScore: result.score,
        score: Math.min(finalScore, 1.0) // Cap at 1.0
      };
    });
    
    // Sort by new score and return top K
    return scoredResults
      .sort((a, b) => b.score - a.score)
      .slice(0, options.topK);
  }
  
  /**
   * Extract keywords from query
   * @param {string} query - User query
   * @returns {Array<string>} Extracted keywords
   */
  extractKeywords(query) {
    // Simple keyword extraction - remove stopwords and extract meaningful terms
    const stopwords = new Set([
      'the', 'and', 'or', 'a', 'an', 'in', 'on', 'at', 'to', 'for', 
      'with', 'by', 'about', 'as', 'how', 'what', 'when', 'where', 
      'who', 'which', 'why', 'can', 'could', 'should', 'would', 'do',
      'does', 'did', 'i', 'you', 'he', 'she', 'we', 'they', 'it'
    ]);
    
    return query
      .toLowerCase()
      .split(/\W+/)
      .filter(word => word.length > 2 && !stopwords.has(word));
  }
  
  /**
   * Fetch document content for results
   * @param {Array<Object>} results - Query results
   * @returns {Array<Object>} Results with content
   */
  async fetchDocumentContent(results) {
    // For results that don't already have content, load it from files
    const resultsWithContent = [];
    
    for (const result of results) {
      let content = null;
      
      // If metadata contains the source file, load content from there
      if (result.metadata && result.metadata.source) {
        try {
          content = await fs.readFile(result.metadata.source, 'utf8');
        } catch (error) {
          console.error(`Error loading content for ${result.id}: ${error.message}`);
          // Try alternative source path
          try {
            const altPath = path.join(this.options.ragDir, result.id);
            content = await fs.readFile(altPath, 'utf8');
          } catch (innerError) {
            // If still can't find content, leave as null
          }
        }
      }
      
      // Add result with content
      resultsWithContent.push({
        ...result,
        content
      });
    }
    
    return resultsWithContent;
  }
  
  /**
   * Merge chunks from the same document
   * @param {Array<Object>} results - Query results with content
   * @returns {Array<Object>} Merged results
   */
  mergeDocumentChunks(results) {
    // Group results by source document
    const documentGroups = {};
    
    results.forEach(result => {
      // Extract document ID without chunk information
      const documentId = result.id.split('_chunk')[0];
      
      if (!documentGroups[documentId]) {
        documentGroups[documentId] = {
          id: documentId,
          score: result.score, // Use highest score
          metadata: { ...result.metadata },
          chunks: []
        };
        
        // Remove chunk-specific metadata
        if (documentGroups[documentId].metadata.chunkIndex !== undefined) {
          delete documentGroups[documentId].metadata.chunkIndex;
          delete documentGroups[documentId].metadata.totalChunks;
        }
      } else if (result.score > documentGroups[documentId].score) {
        // Keep the highest score among chunks
        documentGroups[documentId].score = result.score;
      }
      
      // Add this chunk
      documentGroups[documentId].chunks.push({
        id: result.id,
        content: result.content,
        score: result.score,
        metadata: result.metadata
      });
    });
    
    // Convert groups back to results array
    return Object.values(documentGroups).map(group => {
      // Sort chunks by their original order if available
      group.chunks.sort((a, b) => {
        const indexA = a.metadata.chunkIndex !== undefined ? a.metadata.chunkIndex : 0;
        const indexB = b.metadata.chunkIndex !== undefined ? b.metadata.chunkIndex : 0;
        return indexA - indexB;
      });
      
      // Combine content from all chunks
      const content = group.chunks.map(chunk => chunk.content).join('\n\n');
      
      return {
        id: group.id,
        score: group.score,
        metadata: group.metadata,
        content,
        chunks: group.chunks
      };
    }).sort((a, b) => b.score - a.score);
  }
  
  /**
   * Search for specific information and generate answers
   * @param {string} query - User query
   * @param {Object} options - Search options
   * @returns {Object} Search results with answer
   */
  async search(query, options = {}) {
    const documents = await this.retrieveDocuments(query, options);
    
    return {
      query,
      timestamp: new Date().toISOString(),
      resultsCount: documents.length,
      documents
    };
  }
  
  /**
   * Generate context for an LLM from retrieved documents
   * @param {Array<Object>} documents - Retrieved documents
   * @param {Object} options - Context generation options
   * @returns {string} Formatted context
   */
  generateLlmContext(documents, options = {}) {
    const contextOptions = Object.assign({
      maxContextLength: 10000,
      includeMetadata: true,
      formatType: 'markdown'
    }, options);
    
    let context = '';
    let currentLength = 0;
    
    // Sort documents by score (highest first)
    const sortedDocs = [...documents].sort((a, b) => b.score - a.score);
    
    for (const doc of sortedDocs) {
      // Format document content based on format type
      let docContext = '';
      
      if (contextOptions.formatType === 'markdown') {
        docContext += `## ${doc.id}\n\n`;
        
        if (contextOptions.includeMetadata && doc.metadata) {
          docContext += `**Source:** ${doc.metadata.source || 'Unknown'}\n`;
          docContext += `**System:** ${doc.metadata.system || 'Unknown'}\n`;
          docContext += `**Category:** ${doc.metadata.category || 'General'}\n`;
          docContext += `**Relevance Score:** ${doc.score.toFixed(2)}\n\n`;
        }
        
        docContext += `${doc.content || 'No content available'}\n\n`;
        docContext += `---\n\n`;
      } else {
        // Plain text format
        docContext += `Source: ${doc.id}\n`;
        if (contextOptions.includeMetadata && doc.metadata) {
          docContext += `Source: ${doc.metadata.source || 'Unknown'}\n`;
        }
        docContext += `\n${doc.content || 'No content available'}\n\n`;
        docContext += `----------\n\n`;
      }
      
      // Check if adding this document would exceed max context length
      if (currentLength + docContext.length > contextOptions.maxContextLength) {
        // If we already have some context, stop here
        if (currentLength > 0) break;
        
        // If this is the first document and it's too long, truncate it
        docContext = docContext.substring(0, contextOptions.maxContextLength);
      }
      
      context += docContext;
      currentLength += docContext.length;
    }
    
    return context;
  }
}

module.exports = RagRetriever;

File: src/modules/rag_retriever.rs
use std::path::{Path, PathBuf};
use std::collections::HashMap;
use std::sync::Arc;
use anyhow::{Result, Context, anyhow};
use async_trait::async_trait;
use log::{info, warn, error};
use serde::{Serialize, Deserialize};

/// Vector database adapter trait
#[async_trait]
pub trait VectorDB: Send + Sync {
    /// Find similar documents based on an embedding vector
    async fn find_similar(&self, embedding: &[f32], options: &QueryOptions) -> Result<Vec<DocumentMatch>>;
    
    /// Insert a document with its embedding into the vector database
    async fn insert_document(&self, document: &Document) -> Result<String>;
}

/// Document with metadata and embedding
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Document {
    pub id: Option<String>,
    pub content: String,
    pub metadata: HashMap<String, String>,
    pub embedding: Option<Vec<f32>>,
}

/// Document match with similarity score
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DocumentMatch {
    pub document: Document,
    pub score: f32,
}

/// Query options for vector retrieval
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct QueryOptions {
    pub top_k: usize,
    pub min_score: f32,
    pub filter: Option<HashMap<String, String>>,
}

/// Encoder trait for generating embeddings
#[async_trait]
pub trait Encoder: Send + Sync {
    /// Generate an embedding for the given text
    async fn embed(&self, text: &str) -> Result<Vec<f32>>;
}

/// Options for RAG retriever
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RagOptions {
    pub top_k: usize,
    pub reranking: bool,
    pub min_score: f32,
    pub keyword_boost: bool,
    pub recency_boost: bool,
    pub chunk_merging: bool,
    pub rag_dir: String,
}

impl Default for RagOptions {
    fn default() -> Self {
        Self {
            top_k: 5,
            reranking: true,
            min_score: 0.7,
            keyword_boost: true,
            recency_boost: true,
            chunk_merging: true,
            rag_dir: "rag_knowledge_base".to_string(),
        }
    }
}

/// RAG Retriever
/// Handles query processing and document retrieval for RAG
pub struct RagRetriever {
    vector_db: Arc<dyn VectorDB>,
    encoder: Option<Arc<dyn Encoder>>,
    options: RagOptions,
    initialized: bool,
    ml_analyzer: Option<Arc<dyn MlAnalyzer>>,
}

/// ML Analyzer trait for document embedding generation
#[async_trait]
pub trait MlAnalyzer: Send + Sync {
    async fn generate_document_embedding(&self, text: &str) -> Result<Vec<f32>>;
}

/// FallbackEncoder provides a simple embedding generation when no better option is available
pub struct FallbackEncoder;

#[async_trait]
impl Encoder for FallbackEncoder {
    async fn embed(&self, text: &str) -> Result<Vec<f32>> {
        // Simple hash function to generate pseudo-embeddings
        let hash = text.chars().fold(0i32, |acc, c| {
            ((acc << 5) - acc) + c as i32
        });
        
        // Generate a deterministic but simple vector based on the hash
        let vector: Vec<f32> = (0..512)
            .map(|i| (hash as f32 * (i as f32 + 1.0) * 0.01).sin() * 0.5 + 0.5)
            .collect();
        
        Ok(vector)
    }
}

impl RagRetriever {
    /// Create a new RAG retriever
    pub fn new(
        vector_db: Arc<dyn VectorDB>,
        options: Option<RagOptions>,
        encoder: Option<Arc<dyn Encoder>>,
        ml_analyzer: Option<Arc<dyn MlAnalyzer>>,
    ) -> Self {
        let options = options.unwrap_or_default();
        
        Self {
            vector_db,
            encoder,
            options,
            initialized: false,
            ml_analyzer,
        }
    }
    
    /// Initialize the retriever
    pub async fn initialize(&mut self) -> Result<()> {
        if self.initialized {
            return Ok(());
        }
        
        info!("Initializing RAG retriever...");
        
        if self.encoder.is_none() {
            // If an ML analyzer is available, use it for embedding generation
            if let Some(ml_analyzer) = &self.ml_analyzer {
                info!("Using MLAnalyzer for embedding generation");
                let analyzer_clone = Arc::clone(ml_analyzer);
                
                struct MlAnalyzerEncoder {
                    analyzer: Arc<dyn MlAnalyzer>,
                }
                
                #[async_trait]
                impl Encoder for MlAnalyzerEncoder {
                    async fn embed(&self, text: &str) -> Result<Vec<f32>> {
                        self.analyzer.generate_document_embedding(text).await
                    }
                }
                
                self.encoder = Some(Arc::new(MlAnalyzerEncoder { analyzer: analyzer_clone }) as Arc<dyn Encoder>);
            } else {
                // Use fallback encoder when no alternatives are available
                warn!("Using fallback embedding method - not recommended for production");
                self.encoder = Some(Arc::new(FallbackEncoder) as Arc<dyn Encoder>);
            }
        }
        
        self.initialized = true;
        info!("RAG retriever initialized successfully");
        
        Ok(())
    }
    
    /// Generate embedding for query text
    pub async fn generate_query_embedding(&self, query: &str) -> Result<Vec<f32>> {
        let encoder = self.encoder.as_ref()
            .ok_or_else(|| anyhow!("No encoder available for query embedding generation"))?;
        
        encoder.embed(query).await.context("Error generating query embedding")
    }
    
    /// Build filters for document retrieval
    fn build_filters(&self, filters: Option<HashMap<String, String>>) -> Option<HashMap<String, String>> {
        filters
    }
    
    /// Retrieve relevant documents for a query
    pub async fn retrieve_documents(
        &mut self, 
        query: &str, 
        options: Option<HashMap<String, serde_json::Value>>
    ) -> Result<Vec<DocumentMatch>> {
        self.initialize().await?;
        
        // Merge default options with any provided options
        let query_options = RagOptions::default(); // TODO: Merge with provided options
        
        info!("Retrieving documents for query: \"{}\"", query);
        
        // Generate query embedding
        let query_embedding = self.generate_query_embedding(query).await
            .context("Failed to generate query embedding")?;
        
        // Apply filters
        let filters = self.build_filters(None); // TODO: Extract filters from options
        
        // Retrieve similar documents
        let results = self.vector_db.find_similar(
            &query_embedding, 
            &QueryOptions {
                top_k: if query_options.reranking { 
                    query_options.top_k * 2 
                } else { 
                    query_options.top_k 
                },
                min_score: query_options.min_score,
                filter: filters,
            }
        ).await?;
        
        if results.is_empty() {
            info!("No documents found for query");
            return Ok(vec![]);
        }
        
        info!("Found {} initial matches", results.len());
        
        // Rerank results if enabled
        let final_results = if query_options.reranking {
            self.rerank_results(query, &results, &query_options).await?
        } else {
            // Just take the top K results
            results.into_iter()
                .take(query_options.top_k)
                .collect()
        };
        
        info!("Retrieved {} final documents for query", final_results.len());
        
        Ok(final_results)
    }
    
    /// Rerank search results for improved relevance
    async fn rerank_results(
        &self, 
        query: &str, 
        results: &[DocumentMatch], 
        options: &RagOptions
    ) -> Result<Vec<DocumentMatch>> {
        info!("Reranking {} results", results.len());
        
        // Simple keyword-based reranking for now
        let mut reranked_results = results.to_vec();
        
        if options.keyword_boost {
            // Extract keywords from query (simplified version)
            let query_words: Vec<&str> = query
                .split_whitespace()
                .filter(|word| word.len() > 3) // Only consider words with more than 3 chars
                .collect();
                
            // Boost scores based on keyword presence
            for result in &mut reranked_results {
                for word in &query_words {
                    if result.document.content.to_lowercase().contains(&word.to_lowercase()) {
                        result.score += 0.1; // Simple boost for keyword matches
                    }
                }
            }
        }
        
        // Sort by score (descending)
        reranked_results.sort_by(|a, b| b.score.partial_cmp(&a.score).unwrap_or(std::cmp::Ordering::Equal));
        
        // Apply recency boost if enabled
        if options.recency_boost {
            // TODO: Apply recency boost based on document creation date if available
        }
        
        // Take only the top K results
        let final_results = reranked_results.into_iter()
            .take(options.top_k)
            .collect();
            
        Ok(final_results)
    }
    
    /// Merge chunks for better context (optional)
    fn merge_chunks(&self, documents: Vec<DocumentMatch>) -> Vec<DocumentMatch> {
        // TODO: Implement chunk merging
        documents
    }
}
